src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicates.java

```diff
@@ -0,0 +1,203 @@
+        }
+        if (tagCountMap.get(Utils.MatchType.READ_MISMATCH) != null) {
+            throw new UserException("The reads grouped by the MarkDuplicates key were not the same, indicating that the BAMs are not the same");
+        }
+        if (printSummary) {
+            Utils.MatchType[] values = Utils.MatchType.values();
+            HashSet<Utils.MatchType> matchTypes = Sets.newHashSet(values);
+            System.out.println("##############################");
+            matchTypes.forEach(s -> System.out.println(s + ": " + tagCountMap.getOrDefault(s, 0)));
+        }
+
+        if (crashOnDiff) {
+            for (Utils.MatchType s : Utils.MatchType.values()) {
+                if (s != Utils.MatchType.EQUAL) {
+                    if (tagCountMap.get(s) != null)
+                    throw new UserException("found difference between the two BAMs: " + s + " with count " + tagCountMap.get(s));
+                }
+            }
+        }
+    }
+}
+
+
+class Utils implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    enum MatchType {
+        EQUAL,          // The same number of reads per key, same number of duplicates, and same representative read.
+        SIZE_UNEQUAL,   // Difference number of reads at this key (this means the BAMs are different).
+        READ_MISMATCH,  // The reads at this key are different for the two BAMs (which means the BAMs are different).
+        DIFF_NUM_DUPES, // The reads are the same from each BAM, but there are different number of duplicates marked.
+        DIFFERENT_REPRESENTATIVE_READ,  // The reads are the same and the same number of duplicates, but with a
+                                        // different representative read (usually this is from a tie-breaker).
+    }
+
+    static boolean isNonPrimary(GATKRead read) {
+        return read.isSecondaryAlignment() || read.isSupplementaryAlignment() || read.isUnmapped();
+    }
+
+    // As taken from part of MarkDuplicatesSparkUtils.subkeyForFragment
+    static String groupKey(GATKRead r) {
+        return r.getContig() + "," + ReadUtils.getStrandedUnclippedStart(r) + "," + ReadUtils.readHasMappedMate(r);
+    }
+
+    /**
+     * getDupes returns the metadata about how well the two sets of reads match.
+     * @param f reads from the first bam
+     * @param s reads from the second bam
+     * @param header header (should be the same for both)
+     * @return the type of the match, EQUAL, DIFFERENT_REPRESENTATIVE_READ, etc.
+     */
+    static MatchType getDupes(Iterable<GATKRead> f, Iterable<GATKRead> s, SAMFileHeader header) {
+        List<GATKRead> first = Lists.newArrayList(f);
+        List<GATKRead> second = Lists.newArrayList(s);
+        if (first.size() != second.size()) {
+            return MatchType.SIZE_UNEQUAL;
+        }
+        int size = first.size();
+        if (size > 1) {
```

Comment:
i would not bother with this check - the "time savings" are not worth adding 2 lines of code